src/core/AutoRest.Core/Model/IModelType.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using AutoRest.Core.Utilities;
+using AutoRest.Core.Utilities.Collections;
+using AutoRest.Core.Validation;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace AutoRest.Core.Model
+{
+    /// <summary>
+    /// Defines an interface for client model types.
+    /// </summary>
+    [JsonObject(IsReference = true)]
+    public interface IModelType : IParent, IChild
+    {
+        /// <summary>
+        /// Gets or sets the IModelType name.
+        /// </summary>
+        [Rule(typeof(IsIdentifier))]
+        Fixable<string> Name { get; }
+
+        /// <summary>
+        /// Allows a type to append extra documentation on a property
+        /// </summary>
+        string ExtendedDocumentation { get; }
+
+        /// <summary>
+        /// Allows a type to specify what the default value looks like.
+        /// </summary>
+        string DefaultValue { get; }
+
+        /// <summary>
+        /// Allows a type to specify that it is a constant value
+        /// </summary>
+        bool IsConstant { get; }
+
+        string DeclarationName { get; }
+
+        string ClassName { get; }
+    }
+
+    /// <summary>
+    /// Base implementation for all ModelTypes
+    /// </summary>
+    [JsonObject(IsReference = true)]
+    public abstract class ModelType : IModelType
+    {
+        private CodeModel _codeModel;
+        private readonly Fixable<string> _name = new Fixable<string>();
+
+        /// <summary>
+        /// Gets or sets the IModelType name.
+        /// </summary>
+        public Fixable<string> Name
+        {
+            get{ return _name;}
+            protected set { _name.CopyFrom(value); }
+        }
+
+        public virtual string ClassName => Name.Value;
+
+        public virtual string DeclarationName => Name.Value;
+
+        /// <summary>
+        /// Allows a type to append extra documentation on a property
+        /// </summary>
+        public virtual string ExtendedDocumentation => null;
```

Comment:
I think ExtendedDocumentation is another one that we'll want fixable eventually, but I can't think of the use case right now. I imagine we'll convert them on an as-needed basis?